deps/rabbit/src/rabbit_queue_member_eval.erl

```diff
@@ -0,0 +1,261 @@
+        true ->
+            Ref = erlang:send_after(DefaultInterval, self(), ?EVAL_MSG),
+            {ok, State#state{timer_ref = Ref}};
+        false ->
+            {ok, State, hibernate}
+    end.
+
+handle_call(_Request, _From, State) ->
+    {reply, ok, State}.
+
+handle_cast(eval_trigger, #state{enabled = false} = State) ->
+    {noreply, State, hibernate};
+handle_cast(eval_trigger, #state{timer_ref = OldRef,
+                               short_interval = Time} = State) ->
+    _ = erlang:cancel_timer(OldRef),
+    Ref = erlang:send_after(Time, self(), ?EVAL_MSG),
+    {noreply, State#state{timer_ref = Ref}};
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+handle_info(?EVAL_MSG, #state{default_interval = DefaultInterval,
+                              short_interval = ShortInterval,
+                              target_group_size = TargetSize} = State) ->
+    Res = eval_quorum_queue_membership(TargetSize),
+    NewTimeout = case Res of
+                     noop ->
+                         DefaultInterval;
+                     _ ->
+                         ShortInterval
+                 end,
+    Ref = erlang:send_after(NewTimeout, self(), ?EVAL_MSG),
+    {noreply, State#state{timer_ref = Ref}};
+handle_info(_Info, #state{enabled = false} = State) ->
+    {noreply, State, hibernate};
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%----------------------------------------------------------------------------
+%% Internal functions
+%%----------------------------------------------------------------------------
+
+eval_quorum_queue_membership(TargetSize) ->
+    LocalLeaders = rabbit_amqqueue:list_local_leaders(),
+    ExpectedNodes = rabbit_nodes:list_members(),
+    Running = rabbit_nodes:list_running(),
+    eval_quorum_members(ExpectedNodes, Running, LocalLeaders, TargetSize, noop).
+
+eval_quorum_members(_ExpectedNodes, _Running, [], _TargetSize, Result) ->
+    Result;
+eval_quorum_members(ExpectedNodes, Running, [Q | LocalLeaders], TargetSize, OldResult) ->
+    {ok, Members, {_, LeaderNode}} = ra:members(amqqueue:get_pid(Q)),
+
+    Result =
+    maybe
+        %% Check if Leader is indeed this node
+        LeaderNode ?= node(),
+        %% And that this not is not in maintenance mode
+        true ?= not rabbit_maintenance:is_being_drained_local_read(node()),
+        MemberNodes = [Node || {_, Node} <- Members],
+        Remove = MemberNodes -- ExpectedNodes,
+        case Remove of
+            [] ->
+                add_member(Q, Running, MemberNodes, get_target_size(Q, TargetSize));
+            _ ->
+                remove_members(Q, Remove)
+        end
+    else
+        _ ->
+            noop
+    end,
+    eval_quorum_members(ExpectedNodes, Running, LocalLeaders, TargetSize,
+                        update_result(OldResult, Result)).
+
+add_member(Q, Running, MemberNodes, TargetSize) ->
```

Comment:
This conditionally adds members, so perhaps should be named `maybe_add_member/4`